cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -481,88 +510,84 @@ private void addArchivesToList(List<ArtifactDetail> archives, Row row) throws IO
     }
   }
 
-  // given a row representing plugin metadata, add all plugins in the row to the given map
-  private void addPluginsToMap(Map<ArtifactInfo, List<PluginClass>> map, Row row) throws IOException {
-    // plugin key contains namespace, plugin type, and plugin name
-    PluginKey pluginKey = PluginKey.parse(row.getRow());
-
+  private void addPluginsToMap(Id.Artifact parentArtifactId, Map<ArtifactInfo, List<PluginClass>> map,
+                               Row row) throws IOException {
     // column is the artifact name and version, value is the serialized PluginClass
     for (Map.Entry<byte[], byte[]> column : row.getColumns().entrySet()) {
       ArtifactColumn artifactColumn = ArtifactColumn.parse(column.getKey());
       PluginData pluginData = gson.fromJson(Bytes.toString(column.getValue()), PluginData.class);
 
-      Id.Artifact artifactId =
-        Id.Artifact.from(pluginKey.namespace, artifactColumn.name, artifactColumn.version);
-      ArtifactInfo artifactInfo = new ArtifactInfo(artifactId, pluginData.artifactLocation);
+      // filter out plugins that don't extend this version of the parent artifact
+      if (matches(pluginData, parentArtifactId.getVersion())) {
```

Comment:
Just curious, if the only thing we check if the artifact version is supported in the plugin's support range. What about artifact id (and namespace) match? If we don't need that, they should we be just taking in ArtifactVersion instead of Id.Artifact?